---
title: Data frames and dplyr
---

"Data frame" is R's name for tabular data. We generally want each row in a data frame to represent a unit of observation, and each column to contain a different type of information about the units of observation. Tabular data in this form is called ["tidy data"](http://vita.had.co.nz/papers/tidy-data.html).

This section uses a collection of modern packages collectively known as the [Tidyverse](https://www.tidyverse.org/). R and its predecessor S have a history dating back to 1976. The Tidyverse fixes some dubious design decisions baked into "base R", including having its own slightly improved form of data frame. Sticking to the Tidyverse where possible is generally safer, Tidyverse packages are more willing to generate errors rather than ignore problems.

If the Tidyverse is not already installed, you will need to install it. However on the server we are using today it is already installed.

```{r, eval=FALSE}
# install.packages("tidyverse")
```

We need to load this package in order to use it.

```{r, warning=F, message=F}
library(tidyverse)
```

The `tidyverse` package loads various other packages, setting up a modern R environment. In this section we will be using functions from the `readr` and `dplyr` packages.


R is a language with mini-languages within it that solve specific problem domains. `dplyr` is such a mini-language, a set of "verbs" (functions) that work well together. `dplyr`, with the help of `tidyr` for some more complex operations, provides a way to perform most manipulations on a data frame that you might need.


# Loading data

We will use the `read_csv` function from `readr`. (See also `read.csv` in base R.)

```{r}
gap <- read_csv("gapminder.csv")

gap
```

This is data from [Gapminder](https://www.gapminder.org/) on life expectancy over time in different countries. The "unit of observation" is a country in a particular year.

Note: "tibble" refers to the Tidyverse's improved form of data frame.


# Exploring

The `View` function gives us a spreadsheet-like view of the data frame.

```
View(gap)
```

However understanding this data frame in R should be less a matter of using a graphical interface, and more about using a variety of R functions to interrogate it.

```{r}
nrow(gap)
ncol(gap)
colnames(gap)
class(gap)
typeof(gap)
summary(gap)
```


### Tip {.callout}

A data frame can also be created from vectors, with the `data_frame` function. (See also `data.frame` in base R.) For example:

```{r}
data_frame(foo=c(10,20,30), bar=c("a","b","c"))
```

### Tip {.callout}

A data frame has column names (`colnames`), and base R data frames can also have row names (`rownames`). However the modern convention, which the Tidyverse enforces, is for a data frame to use column names but not row names. Typically a data frame contains a collection of items (rows), each having various properties (columns). If an item has an identifier such as a unique name, this would be given as just another column.


###

The `count` function from `dplyr` can help us understand the structure of this data frame. (See also `table` in base R.) `count` is a little *magical*, we can refer to columns of the data frame directly in the arguments to `count`. 

```{r}
count(gap, year)
count(gap, country)
count(count(gap, country), n)
```

There is data from 142 countries at 12 time points. The data is complete, with no missing values.


# Indexing

Data frames can be subset using `[row,column]` syntax.

```{r}
gap[3,4]
```

Note that this is still wrapped in a data frame. (This is a behaviour specific to Tidyverse data frames.)

Columns can be given by name.

```{r}
gap[3, "lifeExp"]
```

The row or column may be omitted, thereby retrieving the full row or column.

```{r}
gap[3,]
```

Multiple rows or columns may be retrieved using a vector.

```{r}
rows_wanted <- c(1,3,5)
gap[rows_wanted,]
```

Ok, so how do we actually get data out of a data frame?

Under the hood, a data frame is a list of column vectors. This is why `typeof` told us that `gap` was a list. We can use `$` to retrieve columns, as in a list. (Occasionally it is also useful to use `[[ ]]` to retrieve columns, for example if the column name we want is stored in a variable.)

```{r}
head( gap$lifeExp )
head( gap[["lifeExp"]] )
```

To get the lifeExp value of the third row as above, but unwrapped, we can use:

```{r}
gap$lifeExp[3]
```


All of these indexing and access methods can also be used with `<-` to modify values or add new columns. For example, suppose we wanted a GDP column. We can add this to the data frame with:

```{r}
gap$gdp <- gap$gdpPercap * gap$pop
```


## Logical indexing

A method of indexing that we haven't discussed yet is logical indexing. Instead of specifying the row number or numbers that we want, we can give a logical vector which is `TRUE` for the rows we want and `FALSE` otherwise. This can also be used with vectors.

Suppose we want just the data for Australia. `==` is a comparison operator meaning "equal to".

```{r}
is_australia <- gap$country == "Australia"

head(is_australia)
sum(is_australia)
```

We can now grab just those rows of the data frame relating to Australia:

```{r}
gap_australia <- gap[is_australia,]

gap_australia
```

We might also want to know *which* rows our logical vector is `TRUE` for. This is achieved with the `which` function. The result of this can also be used to index the data frame, as we saw earlier.

```{r}
which_australia <- which(is_australia)
which_australia
gap[which_australia,]
```


Comparison operators available are:

* `x == y ` -- "equal to"
* `x != y ` -- "not equal to"
* `x < y  ` -- "less than"
* `x > y  ` -- "greater than"
* `x <= y ` -- "less than or equal to"
* `x >= y ` -- "greater than or equal to"

More complicated conditions can be constructed using logical operators:

* `a & b ` -- "and", true only if both `a` and `b` are true.
* `a | b ` -- "or", true if either `a` or `b` or both are true.
* `! a   ` -- "not" , true if `a` is false, and false if `a` is true.


For example, suppose we wanted to know in which years the life expectancy in Australia was over 75.

```{r}
over_75 <- gap$lifeExp >= 75
is_australia_over_75 <- is_australia & over_75

sum(is_australia_over_75)
gap[is_australia_over_75,]
```

### Challenge {.challenge}

What continents are the countries divided into in this data?

Which countries in Asia had a life expectancy over 75 in 2007?


## A `dplyr` shorthand

The above method is a little laborious. We have to keep mentioning the name of the data frame, and there is a lot of punctuation to keep track of. `dplyr` provides a slightly magical function called `filter` which lets us write more concisely.

```{r}
filter(gap, country == "Australia")
```

In the second argument, we are able to refer to columns of the data frame as though they were variables.



## Different ways to do the same thing

We constructed `gap[is_australia_over_75,]` by creating several variables and then combining them. It is also perfectly possible to do all this in one line:

```{r}
gap[gap$country == "Australia" & gap$lifeExp >= 75,]
```

If you encounter R code that is too difficult to read, it can often be broken down into multiple steps, with intermediate results stored in variables.

However we can't do this with calls to the `filter` function, since it is magic. More precisely, it uses something called ["non-standard evaluation"](http://adv-r.had.co.nz/Computing-on-the-language.html). For example, this doesn't work:

```
is_australia <- year == "Australia"
filter(gap, is_australia)
```



# Basic plotting

Simple plots can be created using the `plot` function.

```{r}
plot(gap_australia$year, gap_australia$lifeExp)
```

However we will see a much more flexible way of plotting in the final section of this workshop.



# Sorting

Data frames can be sorted using the `arrange` function in `dplyr`.

```{r}
arrange(gap, country)
```

The `desc` helper function can be used to arrange in descending order.

```{r}
arrange(gap, desc(country))
```


### Challenge {.challenge}

Which country had the lowest life expectancy in 1952? Which had the highest?




# Summaries

R has a variety of functions for summarizing a vector, including: `sum`, `mean`, `min`, `max`, `median`, `sd`.

```{r}
mean( c(1,2,3,4) )
```

We can use this on the Gapminder data.

```{r}
gap2007 <- filter(gap, year == 2007)
mean(gap2007$lifeExp)
```

(Possibly this should be a `weighted.mean`, as countries have different populations, but let's skip this detail.)

The `summarize` function in `dplyr` allows these to be applied to data frames.

```{r}
summarize(gap2007, mean_lifeExp=mean(lifeExp))
```

So far unremarkable, but `summarize` comes into its own when the `group_by` "adjective" is used. (See also `apply`, `tapply` in base R.)

```{r}
summarize(group_by(gap, year), mean_lifeExp=mean(lifeExp))
```

### Challenge {.challenge}

What is the average of gdpPercap for each continent in 2007?

Advanced: What is the total GDP and total population for each continent in 2007? Therefore, what is the correct GDP per capita for each continent?




# t-test

We will finish this section by demonstrating a t-test as an example of statistical tests available in R.

Has life expectancy increased from 2002 to 2007?

```{r}
gap2002 <- filter(gap, year == 2002)
gap2007 <- filter(gap, year == 2007)

t.test(gap2007$lifeExp, gap2002$lifeExp)
```

This can actually be considered a paired sample t-test. We can specify `paired=TRUE` to `t.test` to perform a paired sample t-test (check this by looking at the help page with `?t.test`). It's important to first check that both data frames are in the same order.

```{r}
all(gap2002$country == gap2007$country)

t.test(gap2007$lifeExp, gap2002$lifeExp, paired=TRUE)
```

When performing a statistical test, it's good practice to visualize the data to make sure there is nothing funny going on.

```{r}
plot(gap2002$lifeExp, gap2007$lifeExp)
abline(0,1)
```


The result of a t-test is actually a value we can manipulate further.

```{r}
result <- t.test(gap2007$lifeExp, gap2002$lifeExp, paired=TRUE)

class(result)
typeof(result)
names(result)

result$p.value
```

In R, a t-test is just another function returning just another type of data, so it can also be a building block.



# Some topics not covered

* Missing data, which R represents as `NA`.

* Factors are a type of vector for categorical data. Factors are similar to character vectors but with an associated ordered set of "levels". See the `factor` function. It may be necessary to convert character vectors into factors, for example to adjust the order of levels when they are displayed in a plot.

* Matrices are similar to data frames, but the columns all contain the same type of data. Conceptually, in a data frame each observation is a row, but in a matrix each observation is a cell in the matrix. See the `matrix` and `as.matrix` functions. These are used in bioinformatics, for example RNA-Seq results summarized as a matrix of read counts associated with genes (rows) and samples (columns).

* Joining data frames together. See functions such as `left_join` and `bind_rows`.

These are covered in our [full day introductory course](https://monashbioinformaticsplatform.github.io/r-intro/).

We also haven't covered programming topics such as writing functions and for-loops. These are covered in our ["more R"](https://monashbioinformaticsplatform.github.io/r-more/) course.

Finally, we have not yet told the full `dplyr` story. For example, we haven't mentioned the pipe [`%>%`](https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html), which is key to writing elegant `dplyr` code. [The R for Data Science](http://r4ds.had.co.nz/transform.html) book provides a more complete introduction to `dplyr`.




